pkg/qrcodegenerator/qrcode/enddevices/enddevices.go

```diff
@@ -0,0 +1,123 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package enddevices
+
+import (
+	"context"
+	"sync"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/qrcodegenerator/qrcode"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+const (
+	serialNumberAttribute = "serial-number"
+	vendorIDAttribute     = "vendor-id"
+	profileIDAttribute    = "profile-id"
+)
+
+var (
+	errCharacter = errors.DefineInvalidArgument("character", "invalid character `{r}`")
+	errNoJoinEUI = errors.DefineFailedPrecondition("no_join_eui", "no JoinEUI")
+	errNoDevEUI  = errors.DefineFailedPrecondition("no_dev_eui", "no DevEUI")
+	errFormat    = errors.DefineInvalidArgument("format", "invalid format")
+)
+
+// Format is a end device QR code format.
+type Format interface {
+	Format() *ttnpb.QRCodeFormat
+	New() Data
+}
+
+// Data represents end device QR code data.
+type Data interface {
+	qrcode.Data
+	Encode(*ttnpb.EndDevice) error
+	// FormatID returns the ID of the format used to parse the QR Code data.
+	FormatID() string
+	// EndDeviceTemplate returns the End Device Template corresponding to the QR code data.
+	EndDeviceTemplate() *ttnpb.EndDeviceTemplate
+}
+
+// Server provides methods for end device QR codes.
+type Server struct {
+	endDeviceFormats sync.Map
```

Comment:
Still, why is this dynamic? When does this change at runtime?
Why isn't this fixed at init time? You can also add one at test init, right?

Apart from options, avoid codebase pollution for tests.

Reply:
> Why isn't this fixed at init time?

It is. https://github.com/TheThingsNetwork/lorawan-stack/blob/4f678e0e27566c47886df58a0ab234ca9845ccdf/pkg/qrcodegenerator/qrcode/enddevices/enddevices.go#L55-L63

This is similar to what we already have. In the current codebase;

- Formats have an option to be updated dynamically; https://github.com/TheThingsNetwork/lorawan-stack/blob/d168a51ebba39db4586790effda102af2fdda536/pkg/qrcode/qrcode.go#L72-L102
- A function is exposed to register formats from outside the package and this is used in the `QRGenerator` tests; https://github.com/TheThingsNetwork/lorawan-stack/blob/d168a51ebba39db4586790effda102af2fdda536/pkg/qrcodegenerator/grpc_test.go#L40

The only change I've made is to use a struct and a `sync.Map`.

Reply:
If this isn't dynamic at runtime, why use a `sync.Map`?

What prevents you from registering the test format at test init time?